File: src/main/java/org/team1/nbe1_3_team01/domain/user/entity/Course.java
package org.team1.nbe1_3_team01.domain.user.entity;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.SQLRestriction;
import org.team1.nbe1_3_team01.domain.board.entity.CourseBoard;
import org.team1.nbe1_3_team01.domain.calendar.entity.CourseSchedule;
import org.team1.nbe1_3_team01.domain.group.entity.Team;

import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@Table(name = "course")
@SQLRestriction("is_delete = false")
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Course {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;

    @Column(columnDefinition = "TINYINT(1)")
    private boolean isDelete;

    @OneToMany(mappedBy = "course")
    private List<User> users = new ArrayList<>();

    @OneToMany(mappedBy = "course")
    private List<Team> teams = new ArrayList<>();

    @OneToMany(mappedBy = "course")
    private List<CourseSchedule> courseSchedules = new ArrayList<>();

    @OneToMany(mappedBy = "course")
    private List<CourseBoard> courseBoards = new ArrayList<>();

    @Builder
    private Course(String name) {
        this.name = name;
        this.isDelete = false;
    }

    public void addUser(User user) {
        this.users.add(user);
    }

    public void addTeam(Team team){
        this.teams.add(team);
    }

    public void addCourseBoard(CourseBoard courseBoard) {
        this.courseBoards.add(courseBoard);
    }

    public void addCourseSchedule(CourseSchedule courseSchedule){
        this.courseSchedules.add(courseSchedule);
    }

    public void updateName(String name) {
        this.name = name;
    }

    public void delete(){
        this.isDelete = true;
    }
}

File: src/main/java/org/team1/nbe1_3_team01/domain/user/entity/Course.kt
package org.team1.nbe1_3_team01.domain.user.entity

import jakarta.persistence.*
import lombok.Builder
import org.hibernate.annotations.SQLRestriction
import org.team1.nbe1_3_team01.domain.board.entity.CourseBoard
import org.team1.nbe1_3_team01.domain.calendar.entity.CourseSchedule
import org.team1.nbe1_3_team01.domain.group.entity.Team

@Entity
@Table(name = "course")
@SQLRestriction("is_delete = false")
class Course private constructor(
    name: String
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null

    var name: String = name
        protected set

    @Column(columnDefinition = "TINYINT(1)")
    var isDelete = false
        protected set

    @OneToMany(mappedBy = "course")
    val users: MutableList<User> = mutableListOf()

    @OneToMany(mappedBy = "course")
    val teams: MutableList<Team> = mutableListOf()

    @OneToMany(mappedBy = "course")
    val courseSchedules: MutableList<CourseSchedule> = mutableListOf()

    @OneToMany(mappedBy = "course")
    val courseBoards: MutableList<CourseBoard> = mutableListOf()

    companion object {
        fun of(
            name: String
        ): Course = Course(
            name = name
        )
    }

    fun addUser(user: User) {
        users.add(user)
    }

    fun addTeam(team: Team) {
        teams.add(team)
    }

    fun addCourseBoard(courseBoard: CourseBoard) {
        courseBoards.add(courseBoard)
    }

    fun addCourseSchedule(courseSchedule: CourseSchedule) {
        courseSchedules.add(courseSchedule)
    }

    fun updateName(name: String) {
        this.name = name
    }

    fun delete() {
        this.isDelete = true
    }
}
